### PR TITLE
Refactor Give Block

### PR DESCRIPTION
**Pull request by johnthepink**
@jbaxleyiii I'm feeling pretty good about the structure of this now. Opening a PR in case anyone has feedback.

### CODE REVIEW

**Comment by jbaxleyiii** [comment]
**[Storybook Hub](https://hub.getstorybook.io/)** is building storybooks for every commit in this PR.

**Latest Storybook: https://e4bb06a5-4529-45c1-a312-120d2bbce1f3.sbook.io/**
<sub>( Visit [here](https://hub.getstorybook.io/app/github.com/NewSpring/Holtzman/branch/refactor-give-block/) for more info )</sub>

> _This comment is automatically added by Storybook Hub_

File: imports/blocks/give/Layout.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    transactionType,
+    scheduleToRecover,
+  } = give;
+
+  if (["loading", "error", "success"].indexOf(state) > -1) {
+    let copiedSchedules;
+
+    switch (state) {
+      case "loading":
+        copiedSchedules = { ...schedules };
```
Comment (jbaxleyiii) [inline imports/blocks/give/Layout.js]: does this persist correctly? Seems like the `let` would be recreated each time?

File: imports/blocks/give/Layout.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            onClick={goToAccounts}
+            schedules={copiedSchedules}
+          />
+        );
+      default:
+        return null;
+    }
+  }
+
+  const Step = FORM_STEPS[step];
```
Comment (jbaxleyiii) [inline imports/blocks/give/Layout.js]: I ❤️ this change

File: imports/blocks/give/Loading.js [hunk 3]
```diff
@@ -1,22 +1,18 @@
-import { PropTypes } from "react";
 import { WindowLoading, Spinner } from "../../components/loading";
 
+const LOADING_HEADER = "We're Processing Your Contribution";
+const LOADING_TEXT = "Please don't close this window while your contribution is being processed.";
```
Comment (jbaxleyiii) [inline imports/blocks/give/Loading.js]: doing constant style text is the way to go! 👍

File: imports/blocks/give/fieldsets/billing/index.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    } else {
+      this.props.save({ billing: { zip: value } });
+    }
+
+    return true;
+  }
+
+  render() {
+    const { billing } = this.props.data;
+    const { states, countries } = this.props;
```
Comment (jbaxleyiii) [inline imports/blocks/give/fieldsets/billing/index.js]: what is the distinction between pulling these off of `props` and not the `this.props[things]` below?

File: imports/blocks/give/fieldsets/payment/index.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    back: PropTypes.func,
+    next: PropTypes.func,
+  }
+
+  state = {
+    save: true,
+  }
+
+  savePayment = () => {
+    this.setState({ save: !this.state.save });
```
Comment (jbaxleyiii) [inline imports/blocks/give/fieldsets/payment/index.js]: @johnthepink should we adjust this to be a function that takes the previous state?

```js
this.setState(({ save }) => ({ save: !save }));
```

File: imports/blocks/give/fieldsets/payment/index.js [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+      accountName: notEmpty,
+      cardNumber: Validate.isCreditCard,
+      expiration: notEmpty,
+      ccv: Validate.isCCV,
+    };
+
+    isValid = validationMap[id](value);
+
+    // special case for intial repaint
+    if ((id === "cardNumber" || id === "routingNumber") && !value) {
```
Comment (jbaxleyiii) [inline imports/blocks/give/fieldsets/payment/index.js]: one line?

File: imports/blocks/give/fieldsets/payment/index.js [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+  // XXX move to layout, but changes to input are needed
+  formatExp = (s, target) => {
+    const save = (adjusted) => {
+      this.saveData(adjusted, target);
+      return adjusted;
+    };
+
+    let current = this.props.data.payment.expiration;
+    if (!current) {
```
Comment (jbaxleyiii) [inline imports/blocks/give/fieldsets/payment/index.js]: one line?

File: imports/blocks/give/fieldsets/payment/index.js [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    const { id } = target;
+
+    const isValid = this.validate(value, target);
+
+    if (isValid) {
+      this.props.save({ payment: { [id]: value } });
+    }
+  }
+
+  // XXX move to layout, but changes to input are needed
```
Comment (jbaxleyiii) [inline imports/blocks/give/fieldsets/payment/index.js]: what changes are needed?

File: imports/blocks/give/fieldsets/payment/index.js [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+    let current = this.props.data.payment.expiration;
+    if (!current) {
+      current = "";
+    }
+
+    const str = `${s}`;
+
+    if (str.length > 5) {
+      return save(str.slice(0, 5));
```
Comment (jbaxleyiii) [inline imports/blocks/give/fieldsets/payment/index.js]: one line?

File: imports/blocks/give/fieldsets/payment/index.js [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+    const isValid = this.validate(value, target);
+
+    if (isValid) {
+      this.props.save({ payment: { [id]: value } });
+    }
+  }
+
+  // XXX move to layout, but changes to input are needed
+  formatExp = (s, target) => {
```
Comment (jbaxleyiii) [inline imports/blocks/give/fieldsets/payment/index.js]: side note, this function is a pain haha

File: imports/blocks/give/fieldsets/personal/Layout.js [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+const NextButton = ({
+  personal,
+  next,
+}) => {
+  const btnClasses = [];
+  let disabled = false;
+  if (
+    personal.email === null || personal.firstName === null ||
+    personal.email === null || personal.campusId === null
```
Comment (jbaxleyiii) [inline imports/blocks/give/fieldsets/personal/Layout.js]: `personal.email` is listed twice

File: imports/blocks/give/fieldsets/shared/ButtonText.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+  let paymentInfo = payment;
+
+  if (!paymentInfo.accountNumber && !paymentInfo.cardNumber) {
+    paymentInfo = { ...savedAccount.payment };
+    paymentInfo.type = "ach";
+  }
+
+  let text = "Give Now";
+
+  if (Object.keys(schedules).length) {
```
Comment (jbaxleyiii) [inline imports/blocks/give/fieldsets/shared/ButtonText.js]: one line this and the ones below?

File: imports/blocks/give/fieldsets/shared/Icon.js [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import { PropTypes } from "react";
+import AccountType from "../../../../components/accountType";
+
+const Icon = ({ cardType }) => (
```
Comment (jbaxleyiii) [inline imports/blocks/give/fieldsets/shared/Icon.js]: just curious, since this just returns `<AccountType>`, why is the extra component needed?

File: imports/blocks/give/fieldsets/shared/cardType.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
+const cardType = (payment, savedAccount) => {
+  if (savedAccount && savedAccount.payment && savedAccount.payment.paymentType) {
+    return savedAccount.payment.paymentType;
+  }
+
+  if (payment.type === "ach") {
```
Comment (jbaxleyiii) [inline imports/blocks/give/fieldsets/shared/cardType.js]: one line

File: imports/blocks/give/fieldsets/shared/isIOS.js [hunk 15]
```diff
@@ -0,0 +1,6 @@
+/* globals cordova */
+const isIOS = () => (
+  typeof cordova !== "undefined" && cordova.platformId === "ios"
```
Comment (jbaxleyiii) [inline imports/blocks/give/fieldsets/shared/isIOS.js]: can we move this to `utils`?

File: imports/blocks/give/fieldsets/shared/monetize.js [hunk 16]
```diff
@@ -0,0 +1,24 @@
+const monetize = (amount, fixed) => {
```
Comment (jbaxleyiii) [inline imports/blocks/give/fieldsets/shared/monetize.js]: this is moved in another PR so we should be able to just pull it in once that's merged 👍

File: imports/blocks/give/index.js [hunk 17]
```diff
@@ -42,7 +38,6 @@ const withCheckout = graphql(CHECKOUT_QUERY, {
   options: { variables: { state: 28, country: 45 } },
```
Comment (jbaxleyiii) [inline imports/blocks/give/index.js]: let's setup the defined values to be the maps needed for inputs as a `props` function. Should be more performant and a good practice to start as it's easier to share

File: imports/blocks/give/index.js [hunk 18]
```diff
@@ -183,118 +159,51 @@ export default class Give extends Component {
...
+      countries = defaultArray,
+      states = defaultArray,
+    } = this.props.data;
 
-    if (!countries) {
-      countries = defaultArray;
-    }
+    const { savedPayments = defaultArray } = this.props.data;
 
+    campuses = campuses.map((x) => ({ label: x.name, value: x.id }));
```
Comment (jbaxleyiii) [inline imports/blocks/give/index.js]: move these to props

File: imports/blocks/give/index.js [hunk 19]
```diff
@@ -183,118 +159,51 @@ export default class Give extends Component {
...
-    }
+  clearData = () => {
+    this.props.dispatch(giveActions.clearData());
+    this.props.dispatch(modal.hide());
+  }
 
-    states = states.map((x) => ({ label: x.name, value: x.value }));
+  render() {
+    let {
+      campuses = defaultArray,
```
Comment (jbaxleyiii) [inline imports/blocks/give/index.js]: this would go away as defaults would be passed from `props` function of `graphql`

**Comment by johnthepink** [comment]
@jbaxleyiii I'm happy to do flow and testing in a different PR, or keep them in this branch. Whatever works best for you.

File: imports/blocks/give/fieldsets/shared/Icon.js (hunk 13 above)
Comment (johnthepink) [inline imports/blocks/give/fieldsets/shared/Icon.js]: There were a lot of places where `<AccountType width="30px" height="21px"` was being used. So my first notion was to abstract it out. I did end up simplifying some of the logic surrounding rendering Icons, which makes the extra component not as necessary. I'm happy to remove it, but I'm not mad at it this way either.

File: imports/blocks/give/fieldsets/shared/monetize.js (hunk 16 above)
Comment (johnthepink) [inline imports/blocks/give/fieldsets/shared/monetize.js]: totally. that's why i just left it here and didn't abstract it all the way out yet.

**Comment by johnthepink** [comment]
@jbaxleyiii pushes fixes for everything.

**Comment by jbaxleyiii** [comment]
@johnthepink I've merged your types into this!

**Comment by jbaxleyiii** [comment]
@johnthepink rebased and merged the testing into this. Aside from the merge conflict, I think this is ready to go. Do you agree?

**Comment by johnthepink** [comment]
@jbaxleyiii eh. looks like i need to fix a date test

**Comment by johnthepink** [comment]
:mole:

**Comment by jbaxleyiii** [comment]
@johnthepink are you okay for this to be merged?
